spec/p2p/reactor/p2p-api.md

```diff
@@ -0,0 +1,276 @@
+The `Switch` instance is provided to every reactor as part of the reactor's
+[registration procedure][reactor-registration].
+The multiple `Peer` instances are provided to every registered reactor whenever
+a [new connection with a peer][reactor-addpeer] is established.
+
+
+## `Switch` API
+
+The [`Switch`][switch-type] is the central component of the p2p layer
+implementation.  It manages all the reactors running in a node and keeps track
+of the connections with peers.
+The table below summarizes the interaction of the standard reactors with the `Switch`:
+
+| `Switch` API method                                     | consensus | block sync | state sync | mempool | evidence  | PEX   |
+|--------------------------------------------|-----------|------------|------------|---------|-----------|-------|
+| `Peers() IPeerSet`                         | x         | x          |            |         |           | x     |
+| `NumPeers() (int, int, int)`               |           | x          |            |         |           | x     |
+| `Broadcast(Envelope) chan bool`            | x         | x          | x          |         |           |       |
+| `MarkPeerAsGood(Peer)`                     | x         |            |            |         |           |       |
+| `StopPeerForError(Peer, interface{})`      | x         | x          | x          | x       | x         | x     |
+| `StopPeerGracefully(Peer)`                 |           |            |            |         |           | x     |
+| `Reactor(string) Reactor`                  |           | x          |            |         |           |       |
+
+The above list is not exhaustive as it does not include all the `Switch` methods
+invoked by the PEX reactor, a special component that should be considered part
+of the p2p layer. This document does not cover the operation of the PEX reactor
+as a connection manager.
+
+### Peers State
+
+The first two methods in the switch API allow reactors to query the state of
+the p2p layer: the set of connected peers.
+
+    func (sw *Switch) Peers() IPeerSet
+
+The `Peers()` method returns the current set of connected peers.
+The returned `IPeerSet` is an immutable concurrency-safe copy of this set.
+Observe that the `Peer` handlers returned by this method were previously
+[added to the reactor][reactor-addpeer] via the `InitPeer(Peer)` method,
+but not yet removed via the `RemovePeer(Peer)` method.
+Thus, a priori, reactors should already have this information.
+
+    func (sw *Switch) NumPeers() (outbound, inbound, dialing int)
+
+The `NumPeers()` method returns the current number of connected peers,
+distinguished between `outbound` and `inbound` peers.
+An `outbound` peer is a peer the node has dialed to, while an `inbound` peer is
+a peer the node has accepted a connection from.
+The third field `dialing` reports the number of peers to which the node is
+currently attempting to connect, so not (yet) connected peers.
```

Comment:
How should this be relevant for a reactor?

Reply:
The PEX reactor uses this field, the others don't. The PEX reactor is a really different component, that in the current version is literally part of the p2p layer.

Reply:
We should really try to find a place for the PEX reactor. In some sense it is a reactor, in some sense it is part of p2p. I think this kind of questions all come from the way PEX is situated.

Reply:
I see. This is one of the first things we will need to fix in the p2p layer when we tackle it. My high-level understanding is that the PEX "reactor" should be using a different interface than the rest, so as not to "pollute" the reactor interface with things that only make sense for PEX.